Refactor complex type handlers and add documentation

IgorD-lab opened the pull request:
Refactor and document the struct type utilities for generating Solidity code from Substrate metadata. It introduces clear type definitions, constants for type mapping, regex patterns for type detection, and improves helper functions for field extraction and type conversion. The code is now more maintainable and easier to extend, with detailed comments explaining each part. Potential bugs marked with "//!".

IgorD-lab committed:
Refactor struct type utilities and add documentation

This commit refactors and documents the struct type utilities for generating Solidity code from Substrate metadata. It introduces clear type definitions, constants for type mapping, regex patterns for type detection, and improves helper functions for field extraction and type conversion. The code is now more maintainable and easier to extend, with detailed comments explaining each part.

IgorD-lab committed:
Improve fixed array parser and add struct doc comments

Refactored the fixed array parser for Rust/Polkadot types to improve delimiter handling and robustness, with expanded documentation and helper functions. Added detailed documentation to the struct type generator explaining its purpose and output for Solidity and SCALE encoding.

IgorD-lab committed:
Add detailed documentation and refactor enum handling

Added documentation for enum variant parsing and Solidity code generation, clarified type mapping and encoder expression functions, and added helper functions for safe encoder naming. Potential bugs and improvements marked with //! and // TODO

Szegoo commented on src/typeDesc/types/enum.ts:
```diff
@@ -72,6 +228,8 @@ export function generateSolidityEnum(typeName: string, json: string | EnumJson):
     .map((v) => {
       if (v.fields.length === 0) {
+        //! Empty string "" vs actual empty bytes is unclear in Solidity
+        //! Use `new bytes(0)` or `hex""` for clarity
```
Use `new bytes(0)` then instead.

Szegoo commented on src/typeDesc/types/enum.ts:
```diff
@@ -117,126 +277,184 @@ ${ctors.replace(/\n/g, '\n    ')}
+  match = normalized.match(TYPE_PATTERNS.OPTION);
+  if (match) return solTypeOf(match[1]); // for function params; tag handled by encoder
 
+  //! Consider logging warning or throwing error for truly unknown types to catch metadata bugs early
```
Warning log should be sufficient.

IgorD-lab committed:
Refactor type mapping and update call encoders

Moved type normalization, regex patterns, and codec name mappings to a new shared typeMapping.ts module to eliminate duplication across enum.ts and struct.ts. Updated enum, struct, and fixedArray generators to use shared helpers. Improved error handling for unknown types and added array length validation. Updated Solidity call encoder contracts to use calldata for struct arguments and removed unused payload structs.
